File: 29Items-Projects/7Algorithms/6-Backtracking/python/knights_tour.py
"""Knight's Tour — visit every square on a chessboard exactly once."""

DX = [-2, -1, 1, 2, 2, 1, -1, -2]
DY = [1, 2, 2, 1, -1, -2, -2, -1]


class KnightsTour:
    def __init__(self, n: int):
        self.n = n
        self.board = [[0] * n for _ in range(n)]

    def solve_and_print(self) -> None:
        print(f"Board: {self.n}×{self.n}\n")
        self.board[0][0] = 1

        if self._backtrack(0, 0, 2):
            self._print_board()
        else:
            print(f"No Knight's Tour exists for n={self.n}")

    def _backtrack(self, row: int, col: int, move: int) -> bool:
        if move > self.n * self.n:
            return True
        for i in range(8):
            nr, nc = row + DX[i], col + DY[i]
            if self._is_valid(nr, nc):
                self.board[nr][nc] = move
                if self._backtrack(nr, nc, move + 1):
                    return True
                self.board[nr][nc] = 0
        return False

    def _is_valid(self, r: int, c: int) -> bool:
        return 0 <= r < self.n and 0 <= c < self.n and self.board[r][c] == 0

    def _print_board(self) -> None:
        width = len(str(self.n * self.n)) + 1
        for row in self.board:
            print("".join(f"{v:{width}d}" for v in row))

File: 29Items-Projects/7Algorithms/6-Backtracking/java/src/knightstour/KnightsTour.java
package knightstour;

public class KnightsTour {

    private final int n;
    private final int[][] board;
    private final int[] dx = {-2, -1, 1, 2, 2, 1, -1, -2};
    private final int[] dy = {1, 2, 2, 1, -1, -2, -2, -1};
    private boolean solved = false;

    public KnightsTour(int n) {
        this.n = n;
        this.board = new int[n][n];
    }

    public void solveAndPrint() {
        System.out.printf("Board: %dx%d%n%n", n, n);
        board[0][0] = 1;
        solved = backtrack(0, 0, 2);

        if (solved) {
            printBoard();
        } else {
            System.out.println("No Knight's Tour exists for n=" + n);
        }
    }

    private boolean backtrack(int row, int col, int moveNum) {
        if (moveNum > n * n) return true;

        for (int i = 0; i < 8; i++) {
            int nr = row + dx[i];
            int nc = col + dy[i];
            if (isValid(nr, nc)) {
                board[nr][nc] = moveNum;
                if (backtrack(nr, nc, moveNum + 1)) return true;
                board[nr][nc] = 0;
            }
        }
        return false;
    }

    private boolean isValid(int r, int c) {
        return r >= 0 && r < n && c >= 0 && c < n && board[r][c] == 0;
    }

    private void printBoard() {
        int width = String.valueOf(n * n).length() + 1;
        for (int r = 0; r < n; r++) {
            for (int c = 0; c < n; c++) {
                System.out.printf("%" + width + "d", board[r][c]);
            }
            System.out.println();
        }
    }
}
